File: 1706-min-cost-to-connect-all-points/min-cost-to-connect-all-points.py
from collections import defaultdict
from typing import List
from heapq import heappop,heappush
class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        N = len(points)
        cost = 0
        adj = defaultdict(list)
        for i in range(N):
            currX,currY = points[i]
            for j in range(i+1,N):
                nextX,nextY = points[j]
                dist = abs(currX-nextX)+abs(currY-nextY)
                adj[(currX,currY)].append([nextX,nextY,dist])
                adj[(nextX,nextY)].append([currX,currY,dist])
        minHeap = []
        p1 = tuple(points.pop(0))
        visit = defaultdict(bool)
        visit[p1] = True
        for neX,neY,d in adj[p1]: heappush(minHeap,(d,neX,neY))
        while minHeap:
            currCost,curX,curY = heappop(minHeap)
            if visit[(curX,curY)]: continue
            visit[(curX,curY)] = True
            cost+=currCost
            for neighX,neighY,DIST in adj[(curX,curY)]:
                if not visit[(neighX,neighY)]: heappush(minHeap,(DIST,neighX,neighY))
        return cost

File: 1706-min-cost-to-connect-all-points/min-cost-to-connect-all-points.cpp
#include <bits/stdc++.h>
using namespace std;
struct Node {
    int d,node;
    bool operator>(const Node& other) const {
        return d > other.d;
    }
};
class Solution {
public:
    int minCostConnectPoints(vector<vector<int>>& points) {
        int cost = 0;
        int N = points.size();
        unordered_map<int,vector<Node>> adj;
        for (int i = 0; i<N;i++) {
            int currX = points[i][0],currY = points[i][1];
            for (int j = i+1;j<N;j++) {
                int nextX = points[j][0],nextY = points[j][1];
                int dist = abs(currX-nextX) + abs(currY-nextY);
                adj[i].push_back({dist,j});
                adj[j].push_back({dist,i});
            }
        }
        priority_queue<Node,vector<Node>,greater<Node>> minHeap;
        unordered_map<int,bool> visit;
        visit[0] = true;
        for (auto [d,neigh]: adj[0]){minHeap.push({d,neigh});}
        while (!minHeap.empty()) {
            Node currN = minHeap.top();minHeap.pop();
            if (visit[currN.node]) continue;
            visit[currN.node] = true;
            cost+=currN.d;
            for (auto [D,ne]: adj[currN.node]) {
                if (visit[ne]!=true) { minHeap.push({D,ne}); }
            }
        }
        return cost;
    }
};